zmq_addon.hpp

```diff
@@ -154,7 +155,122 @@ send_multipart(socket_ref s, Range &&msgs, send_flags flags = send_flags::none)
+   size less than 255 bytes will have a single byte size value.
+   Larger parts will have a five byte size value with the first byte
+   set to 0xFF and the remaining four bytes holding the size of the
+   part's data.
+*/
+template<class Range
+#ifndef ZMQ_CPP11_PARTIAL
+         ,
+         typename = typename std::enable_if<
+           detail::is_range<Range>::value
+           && (std::is_same<detail::range_value_t<Range>, message_t>::value
+               || detail::is_buffer<detail::range_value_t<Range>>::value)>::type
 #endif
+         >
+message_t encode(const Range &parts)
+{
+    size_t mmsg_size = 0;
+
+    // First pass check sizes
+    for (const message_t &part : parts) {
+        size_t part_size = part.size();
+        if (part_size > 0xFFFFFFFF) {
+            // Size value must fit into uint32_t.
+            throw std::range_error("Invalid size, message part too large");
+        }
+        size_t count_size = 5;
+        if (part_size < 255) {
+            count_size = 1;
+        }
+        mmsg_size += part_size + count_size;
+    }
+
+    message_t encoded(mmsg_size);
+    unsigned char *buf = encoded.data<unsigned char>();
+    for (const message_t &part : parts) {
+        uint32_t part_size = part.size();
+        const unsigned char *part_data = part.data<const unsigned char>();
+
+        // small part
+        if (part_size < 255) {
+            *buf++ = (unsigned char) part_size;
+            memcpy(buf, part_data, part_size);
+            buf += part_size;
+            continue;
+        }
+
+        // big part
+        *buf++ = 0xFF;
+        *buf++ = (part_size >> 24) & 255;
+        *buf++ = (part_size >> 16) & 255;
+        *buf++ = (part_size >> 8) & 255;
+        *buf++ = part_size & 255;
+        memcpy(buf, part_data, part_size);
+        buf += part_size;
+    }
+    return encoded;
+}
+
+/*  Decode an encoded message to multiple parts.
+
+    The given output iterator must be a ForwardIterator to a container
+    holding zmq::message_t such as a zmq::multipart_t or various STL
+    containers.
+
+    Returns the ForwardIterator advanced once past the last decoded
+    part.
+
+    Throws: a std::out_of_range is thrown if the encoded part sizes
+    lead to exceeding the message data bounds.
+
+    The decoding assumes the message is encoded in the manner
+    performed by zmq::encode().
+ */
+template<class OutputIt> OutputIt decode(const message_t &encoded, OutputIt out)
+{
+    const unsigned char *source = encoded.data<unsigned char>();
+    const unsigned char *const limit = source + encoded.size();
+
+    while (source < limit) {
+        size_t part_size = *source++;
+        if (part_size == 255) {
+            if (source > limit - 4) {
+                throw std::out_of_range(
+                  "Malformed encoding, overflow in reading size");
+            }
+            part_size =
+              (source[0] << 24) + (source[1] << 16) + (source[2] << 8) + source[3];
```

Comment:
The types here are `unsigned char` and `int`, so I think there could be overflow if `source[0]` is big enough. There should probably by casts to `uint32_t`.

Reply:
Indeed!  I confirm if `source[0]` is, eg `0xff` this will not give the expected value.